spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/SparkMLPredictor.java

```diff
@@ -0,0 +1,153 @@
+      this.fileSetName = fileSetName;
+      this.path = path;
+      this.featureFieldsToInclude = featureFieldsToInclude;
+      this.featureFieldsToExclude = featureFieldsToExclude;
+      this.predictionField = predictionField;
+    }
+
+    public void validate(Schema inputSchema) {
+      SparkUtils.validateConfigParameters(inputSchema, featureFieldsToInclude, featureFieldsToExclude, predictionField,
+                                          null);
+    }
+  }
+
+  public SparkMLPredictor(MLPredictorConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+    Preconditions.checkArgument(inputSchema != null, "Input Schema must be a known constant.");
+    config.validate(inputSchema);
+    stageConfigurer.setOutputSchema(SparkUtils.getOutputSchema(inputSchema, config.predictionField));
+  }
+
+  @Override
+  public void initialize(SparkExecutionPluginContext context) throws Exception {
+    FileSet fileSet = context.getDataset(config.fileSetName);
+    Location modelLocation = fileSet.getBaseLocation().append(config.path);
+
+    if (!modelLocation.exists()) {
+      throw new IllegalArgumentException("Failed to find model to use for classification. " +
+                                           "Location does not exist: " + modelLocation, null);
+    }
+    initialize(context.getSparkContext().sc(), modelLocation.toURI().getPath());
+  }
+
+  public abstract void initialize(SparkContext context, String modelPath);
+
+  public abstract double predict(Vector vector);
+
+  @Override
+  public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext context,
+                                             JavaRDD<StructuredRecord> input) throws Exception {
+    if (input.isEmpty()) {
+      return context.getSparkContext().emptyRDD();
+    }
+    Schema inputSchema = input.first().getSchema();
+    outputSchema = (outputSchema != null) ? outputSchema :
+      SparkUtils.getOutputSchema(inputSchema, config.predictionField);
+    final Map<String, Integer> featuresList = SparkUtils.getFeatureList(inputSchema,
+                                                                        config.featureFieldsToInclude,
+                                                                        config.featureFieldsToExclude,
+                                                                        config.predictionField);
+    JavaRDD<StructuredRecord> output = input.map(new Function<StructuredRecord, StructuredRecord>() {
+      @Override
+      public StructuredRecord call(StructuredRecord record) throws Exception {
+        List<Double> featureList = new ArrayList<>();
+        List<Integer> featureIndex = new ArrayList<>();
+        int counter = 0;
```

Comment:
size can be found by just doing featuresList.size(). The featureList and featureIndex lists can then be arrays. That way we don't have to turn the lists into arrays and save two traversals